// @scaleflex/ui/core/check-box/check-box.component.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["size", "type", "checked", "onChange", "checkBoxProps", "readOnly", "disabled"];
import React from 'react';
import PT from 'prop-types';
import Tick from '@scaleflex/icons/tick';
import Intermediate from '@scaleflex/icons/intermediate';
import { lightPalette } from '../../theme/roots/palette';
import { Color } from '../../utils/types/palette';
import { intrinsicComponent, objectValues } from '../../utils/functions';
import Styled from './check-box.styles';
import { Size, Type } from './types';
import { getCheckboxIconSize } from './check-box.utils';
var CheckBox = intrinsicComponent(function (_ref, ref) {
  var _ref$size = _ref.size,
    size = _ref$size === void 0 ? Size.Md : _ref$size,
    _ref$type = _ref.type,
    type = _ref$type === void 0 ? Type.Checkbox : _ref$type,
    checked = _ref.checked,
    onChange = _ref.onChange,
    checkBoxProps = _ref.checkBoxProps,
    readOnly = _ref.readOnly,
    disabled = _ref.disabled,
    rest = _objectWithoutProperties(_ref, _excluded);
  var getIcon = function getIcon(props) {
    switch (type) {
      case Type.Checkbox:
        return /*#__PURE__*/React.createElement(Tick, props);
      case Type.Intermediate:
        return /*#__PURE__*/React.createElement(Intermediate, props);
      default:
        return /*#__PURE__*/React.createElement(Tick, props);
    }
  };
  return /*#__PURE__*/React.createElement(Styled.CheckBox, _extends({
    size: size,
    disabled: Boolean(disabled),
    checked: Boolean(checked),
    ref: ref
  }, rest), /*#__PURE__*/React.createElement(Styled.Input, _extends({
    checked: checked,
    disabled: disabled,
    size: size,
    onChange: readOnly || disabled ? undefined : onChange
  }, checkBoxProps)), checked ? getIcon({
    size: getCheckboxIconSize(size),
    color: lightPalette[Color.ButtonPrimaryText]
  }) : /*#__PURE__*/React.createElement(Styled.UnCheckedIcon, null, getIcon({
    size: getCheckboxIconSize(size),
    color: lightPalette[Color.Accent_4_0_Opacity]
  })));
});
CheckBox.defaultProps = {
  checked: false,
  size: Size.Md
};
CheckBox.propTypes = {
  checked: PT.bool,
  onChange: PT.func,
  checkBoxProps: PT.object,
  readOnly: PT.bool,
  disabled: PT.bool,
  size: PT.oneOf(objectValues(Size)),
  type: PT.oneOf(objectValues(Type))
};
export default CheckBox;
