// @scaleflex/ui/core/popup-content/popup-content.component.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["onClose", "message", "status"];
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
import React from 'react';
import PT from 'prop-types';
import { intrinsicComponent } from '../../utils/functions';
import PopupStatus from '../popup-status';
import CrossButton from '../cross-button/cross-button.component';
import { defaultProps as popupStatusDefaultProps, propTypes as popupStatusPropTypes } from '../popup-status/popup-status.component';
import Styled from './popup-content.styles';
var PopupContent = intrinsicComponent(function (_ref, ref) {
  var onClose = _ref.onClose,
    message = _ref.message,
    status = _ref.status,
    rest = _objectWithoutProperties(_ref, _excluded);
  return /*#__PURE__*/React.createElement(Styled.PopupContent, _extends({}, rest, {
    ref: ref
  }), /*#__PURE__*/React.createElement(Styled.PopupStatus, _extends({
    status: status,
    message: message
  }, rest), /*#__PURE__*/React.createElement(PopupStatus, {
    status: status
  })), /*#__PURE__*/React.createElement(Styled.LabelWrapper, null, /*#__PURE__*/React.createElement(Styled.Label, null, message)), /*#__PURE__*/React.createElement(Styled.CloseWrapper, null, /*#__PURE__*/React.createElement(CrossButton, {
    size: "sm",
    onClick: onClose
  })));
});
export var defaultProps = _objectSpread({}, popupStatusDefaultProps);
PopupContent.defaultProps = defaultProps;
export var propTypes = _objectSpread(_objectSpread({}, popupStatusPropTypes), {}, {
  message: PT.node.isRequired,
  onClose: PT.func
});
PopupContent.propTypes = propTypes;
export default PopupContent;
