// @scaleflex/ui/core/calendar/calendar.styles.js
import styled, { css, keyframes } from 'styled-components';
import { generateClassNames, applyDisplayNames } from '../../utils/functions';
import { FontVariant as FV } from '../../utils/types/typography';
import { Color as PColor } from '../../utils/types/palette';
import { getCalendarButtonsBackgroundColor, getCalendarButtonsColor, getDatePickerDaysColor } from './calendar.utils';
var baseClassName = 'Calendar';
var rightDatePickerKeyframes = /*#__PURE__*/keyframes(["0%{transform:translateX(-100%)}100%{transform:translateX(0)}"]);
var rightDatePickerAnimation = /*#__PURE__*/css(["animation:", " 300ms ease-in-out;"], rightDatePickerKeyframes);
var leftDatePickerKeyframes = /*#__PURE__*/keyframes(["0%{transform:translateX(100%)}100%{transform:translateX(0)}"]);
var leftDatePickerAnimation = /*#__PURE__*/css(["animation:", " 300ms ease-in-out;"], leftDatePickerKeyframes);
var Calendar = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'root')
}).withConfig({
  componentId: "sc-10lx57a-0"
})(function (_ref) {
  var palette = _ref.theme.palette;
  return css(["position:absolute;width:224px;top:4px;background:", ";box-shadow:0px 2px 6px ", ";border-radius:2px;overflow:hidden;padding:14px 12px 12px 12px;z-index:1111;user-select:none;"], palette[PColor.BackgroundStateless], palette[PColor.LargeShadow]);
});
var DatePickerCalendar = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'date-picker-calendar')
}).withConfig({
  componentId: "sc-10lx57a-1"
})(function () {
  return css(["width:100%;height:100%;"]);
});
var DatePickerCalendarHead = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'date-picker-calendar-head')
}).withConfig({
  componentId: "sc-10lx57a-2"
})(function () {
  return css(["display:flex;justify-content:space-between;height:24px;width:100%;margin-bottom:5px;padding:2px 4px;align-items:center;"]);
});
var DatePickerCalendarBody = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'date-picker-calendar-body')
}).withConfig({
  componentId: "sc-10lx57a-3"
})(function (_ref2) {
  var _ref2$isNextMonth = _ref2.isNextMonth,
    isNextMonth = _ref2$isNextMonth === void 0 ? false : _ref2$isNextMonth,
    _ref2$isPrevMonth = _ref2.isPrevMonth,
    isPrevMonth = _ref2$isPrevMonth === void 0 ? false : _ref2$isPrevMonth;
  return css(["position:relative;display:block;height:174px;width:100%;float:left;", " ", ""], isNextMonth && rightDatePickerAnimation, isPrevMonth && leftDatePickerAnimation);
});
var DatePickerHeadDay = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'date-picker-head-day')
}).withConfig({
  componentId: "sc-10lx57a-4"
})(function (_ref3) {
  var _ref3$theme = _ref3.theme,
    palette = _ref3$theme.palette,
    font = _ref3$theme.typography.font;
  return css(["display:flex;align-items:center;width:calc(100% / 7);height:20px;color:", ";text-align:center;justify-content:center;", ""], palette[PColor.TextSecondary], font[FV.LabelSmall]);
});
var MonthDatePickerWrapper = /*#__PURE__*/styled(Calendar).attrs({
  className: generateClassNames(baseClassName, 'month-date-picker-wrapper')
}).withConfig({
  componentId: "sc-10lx57a-5"
})(function (_ref4) {
  var _ref4$open = _ref4.open,
    open = _ref4$open === void 0 ? false : _ref4$open;
  return css(["top:0;left:0;box-shadow:unset;transform:", ";transition:transform 0.4s ease-in-out;"], open ? 'translateY(0)' : 'translateY(-100%)');
});
var MonthButtonsWrapper = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'month-buttons-wrapper')
}).withConfig({
  componentId: "sc-10lx57a-6"
})(function () {
  return css(["display:flex;flex-wrap:wrap;justify-content:center;padding:3px 0 28px 0;width:100%;height:244px;"]);
});
var MonthButtons = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'month-buttons')
}).withConfig({
  componentId: "sc-10lx57a-7"
})(function (_ref5) {
  var _ref5$isYearChanged = _ref5.isYearChanged,
    isYearChanged = _ref5$isYearChanged === void 0 ? false : _ref5$isYearChanged,
    _ref5$isMonthChanged = _ref5.isMonthChanged,
    isMonthChanged = _ref5$isMonthChanged === void 0 ? false : _ref5$isMonthChanged,
    _ref5$isDisabled = _ref5.isDisabled,
    isDisabled = _ref5$isDisabled === void 0 ? false : _ref5$isDisabled,
    palette = _ref5.theme.palette;
  return css(["width:calc(100% / 3);display:flex;justify-content:center;align-items:center;max-width:58.67px;height:24px;max-height:24px;text-align:center;min-height:unset;border-radius:2px;cursor:", ";pointer-events:", ";font-size:13px;line-height:14px;&:hover{", "}background-color:", ";color:", ";"], isDisabled ? 'auto' : 'pointer', isDisabled ? 'none' : '', !isMonthChanged && !isYearChanged && !isDisabled && css(["background-color:transparent;border:1px solid ", ";box-sizing:border-box;line-height:22px;"], palette[PColor.AccentPrimaryHover]), getCalendarButtonsBackgroundColor(isYearChanged, isMonthChanged, isDisabled), getCalendarButtonsColor(isYearChanged, isMonthChanged, isDisabled));
});
var CalendarBody = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'date-picker-body')
}).withConfig({
  componentId: "sc-10lx57a-8"
})(function () {
  return css(["height:174px;width:100%;"]);
});
var DatePickerDays = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'date-picker-days')
}).withConfig({
  componentId: "sc-10lx57a-9"
})(function () {
  return css(["position:relative;display:block;float:left;width:calc(100% / 7);height:30px;"]);
});
var DatePickerDay = /*#__PURE__*/styled.span.attrs({
  className: generateClassNames(baseClassName, 'date-picker-day')
}).withConfig({
  componentId: "sc-10lx57a-10"
})(function (_ref6) {
  var day = _ref6.day,
    _ref6$isSelectedDay = _ref6.isSelectedDay,
    isSelectedDay = _ref6$isSelectedDay === void 0 ? false : _ref6$isSelectedDay,
    _ref6$isDisabled = _ref6.isDisabled,
    isDisabled = _ref6$isDisabled === void 0 ? false : _ref6$isDisabled,
    palette = _ref6.theme.palette;
  return css(["width:24.57px;height:24px;margin-top:-15px;left:50%;top:50%;font-weight:400;border-radius:2px;line-height:23px;color:", ";pointer-events:", ";cursor:pointer;transition:all 100ms ease-out;background-color:", ";&:hover{", "}"], getDatePickerDaysColor(day, isSelectedDay, isDisabled), day.month !== 0 || isDisabled ? 'none' : '', isSelectedDay && day.month === 0 ? palette[PColor.AccentStateless] : palette[PColor.BackgroundStateless], !isSelectedDay && !isDisabled && css(["background-color:transparent;border:1px solid ", ";box-sizing:border-box;line-height:21px;"], palette[PColor.AccentPrimaryHover]));
});
var DatePickerDayContainer = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'date-picker-day-container')
}).withConfig({
  componentId: "sc-10lx57a-11"
})(function (_ref7) {
  var palette = _ref7.theme.palette;
  return css(["position:relative;display:flex;justify-content:center;align-items:center;width:38.5px;height:35px;font-size:13px;line-height:14px;color:", ";text-align:center;"], palette[PColor.TextPrimary]);
});
var HeaderWrapper = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'header-wrapper')
}).withConfig({
  componentId: "sc-10lx57a-12"
})(function () {
  return css(["display:flex;position:relative;align-items:center;width:100%;justify-content:center;margin-bottom:25px;height:16px;"]);
});
var HeaderBody = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'header-body')
}).withConfig({
  componentId: "sc-10lx57a-13"
})(function () {
  return css(["display:flex;align-items:center;column-gap:4px;max-width:93px;height:100%;"]);
});
var YearsHeaderBody = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'years-header-body')
}).withConfig({
  componentId: "sc-10lx57a-14"
})(function () {
  return css(["display:flex;align-items:center;justify-content:center;column-gap:4px;width:fit-content;height:100%;cursor:pointer;"]);
});
var HeaderBodyYear = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'header-body-year')
}).withConfig({
  componentId: "sc-10lx57a-15"
})(function (_ref8) {
  var _ref8$theme = _ref8.theme,
    palette = _ref8$theme.palette,
    font = _ref8$theme.typography.font;
  return css(["width:100%;color:", ";font-weight:normal;text-align:center;cursor:pointer;", ""], palette[PColor.TextPrimary], font[FV.InputSmUp]);
});
var HeaderBodyMonth = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'header-body-month')
}).withConfig({
  componentId: "sc-10lx57a-16"
})(function (_ref9) {
  var _ref9$theme = _ref9.theme,
    palette = _ref9$theme.palette,
    font = _ref9$theme.typography.font;
  return css(["width:100%;font-weight:normal;text-align:center;color:", ";cursor:pointer;", ""], palette[PColor.TextPrimary], font[FV.InputSmUp]);
});
var HeaderLeftArrows = /*#__PURE__*/styled.span.attrs({
  className: generateClassNames(baseClassName, 'header-left-arrow')
}).withConfig({
  componentId: "sc-10lx57a-17"
})(function (_ref10) {
  var palette = _ref10.theme.palette,
    _ref10$isDisabled = _ref10.isDisabled,
    isDisabled = _ref10$isDisabled === void 0 ? false : _ref10$isDisabled;
  return css(["position:absolute;left:0;top:0;color:", ";cursor:", ";pointer-events:", ";"], isDisabled ? palette[PColor.TextPlaceholder] : palette[PColor.IconsPrimary], isDisabled ? 'auto' : 'pointer', isDisabled ? 'none' : '');
});
var HeaderLeftArrow = /*#__PURE__*/styled.span.attrs({
  className: generateClassNames(baseClassName, 'header-left-arrow')
}).withConfig({
  componentId: "sc-10lx57a-18"
})(function (_ref11) {
  var palette = _ref11.theme.palette,
    isDisabled = _ref11.isDisabled;
  return css(["position:absolute;left:16.37px;top:0;color:", ";cursor:", ";pointer-events:", ";"], isDisabled ? palette[PColor.TextPlaceholder] : palette[PColor.IconsPrimary], isDisabled ? 'auto' : 'pointer', isDisabled ? 'none' : '');
});
var MonthsHeaderLeftArrow = /*#__PURE__*/styled.span.attrs({
  className: generateClassNames(baseClassName, 'month-header-left-arrow')
}).withConfig({
  componentId: "sc-10lx57a-19"
})(function (_ref12) {
  var palette = _ref12.theme.palette,
    isDisabled = _ref12.isDisabled;
  return css(["position:absolute;color:", ";left:0;top:0;color:", ";cursor:", ";pointer-events:", ";"], palette[PColor.IconsPrimary], isDisabled ? palette[PColor.TextPlaceholder] : palette[PColor.IconsPrimary], isDisabled ? 'auto' : 'pointer', isDisabled ? 'none' : '');
});
var HeaderRightArrow = /*#__PURE__*/styled.span.attrs({
  className: generateClassNames(baseClassName, 'header-right-arrow')
}).withConfig({
  componentId: "sc-10lx57a-20"
})(function (_ref13) {
  var palette = _ref13.theme.palette,
    isDisabled = _ref13.isDisabled;
  return css(["position:absolute;right:16.37px;top:0;color:", ";cursor:", ";pointer-events:", ";"], isDisabled ? palette[PColor.TextPlaceholder] : palette[PColor.IconsPrimary], isDisabled ? 'auto' : 'pointer', isDisabled ? 'none' : '');
});
var MonthsHeaderRightArrow = /*#__PURE__*/styled.span.attrs({
  className: generateClassNames(baseClassName, 'month-header-right-arrow')
}).withConfig({
  componentId: "sc-10lx57a-21"
})(function (_ref14) {
  var palette = _ref14.theme.palette,
    isDisabled = _ref14.isDisabled;
  return css(["position:absolute;right:0;color:", ";cursor:", ";pointer-events:", ";"], isDisabled ? palette[PColor.TextPlaceholder] : palette[PColor.IconsPrimary], isDisabled ? 'auto' : 'pointer', isDisabled ? 'none' : '');
});
var HeaderRightArrows = /*#__PURE__*/styled.span.attrs({
  className: generateClassNames(baseClassName, 'header-right-arrows')
}).withConfig({
  componentId: "sc-10lx57a-22"
})(function (_ref15) {
  var palette = _ref15.theme.palette,
    _ref15$isDisabled = _ref15.isDisabled,
    isDisabled = _ref15$isDisabled === void 0 ? false : _ref15$isDisabled;
  return css(["position:absolute;right:0;color:", ";top:0;cursor:", ";pointer-events:", ";;"], isDisabled ? palette[PColor.TextPlaceholder] : palette[PColor.IconsPrimary], isDisabled ? 'auto' : 'pointer', isDisabled ? 'none' : '');
});
var ButtonWrapper = /*#__PURE__*/styled.span.attrs({
  className: generateClassNames(baseClassName, 'button-wrapper')
}).withConfig({
  componentId: "sc-10lx57a-23"
})(function () {
  return css(["display:flex;justify-content:flex-end;align-items:center;column-gap:4px;width:100%;height:24px;padding-top:12px;.SfxButton-root{max-height:24px;}"]);
});
var Styled = applyDisplayNames({
  Calendar: Calendar,
  MonthDatePickerWrapper: MonthDatePickerWrapper,
  MonthButtonsWrapper: MonthButtonsWrapper,
  MonthButtons: MonthButtons,
  CalendarBody: CalendarBody,
  DatePickerDays: DatePickerDays,
  DatePickerDay: DatePickerDay,
  DatePickerDayContainer: DatePickerDayContainer,
  DatePickerCalendarHead: DatePickerCalendarHead,
  DatePickerCalendarBody: DatePickerCalendarBody,
  DatePickerHeadDay: DatePickerHeadDay,
  DatePickerCalendar: DatePickerCalendar,
  HeaderWrapper: HeaderWrapper,
  HeaderBody: HeaderBody,
  YearsHeaderBody: YearsHeaderBody,
  ButtonWrapper: ButtonWrapper,
  HeaderBodyYear: HeaderBodyYear,
  HeaderBodyMonth: HeaderBodyMonth,
  HeaderLeftArrows: HeaderLeftArrows,
  MonthsHeaderLeftArrow: MonthsHeaderLeftArrow,
  HeaderRightArrows: HeaderRightArrows,
  HeaderRightArrow: HeaderRightArrow,
  MonthsHeaderRightArrow: MonthsHeaderRightArrow,
  HeaderLeftArrow: HeaderLeftArrow
});
export default Styled;
