// @scaleflex/ui/theme/roots/typography/entity/default-typography.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
var _labelFontVariants, _textFontVariants, _buttonFontVariants, _linkFontVariants, _inputFontVariants, _titleFontVariants;
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
/* eslint-disable sonarjs/no-duplicate-string */
import { FontVariant, FontWeight } from '../../../../utils/types/typography';
var fontFamily = '"Roboto"';
var defaultFontValue = {
  fontFamily: fontFamily,
  fontStyle: 'normal',
  fontWeight: FontWeight.Regular
};
var labelFontVariants = (_labelFontVariants = {}, _defineProperty(_labelFontVariants, FontVariant.LabelExtraSmall, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontSize: '10px',
  lineHeight: '12px'
})), _defineProperty(_labelFontVariants, FontVariant.LabelExtraSmallEmphasis, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontWeight: 'Bold',
  fontSize: '10px',
  lineHeight: '12px'
})), _defineProperty(_labelFontVariants, FontVariant.LabelExtraSmallUp, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontSize: '11px',
  lineHeight: '12px'
})), _defineProperty(_labelFontVariants, FontVariant.LabelSmall, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontSize: '12px',
  lineHeight: '14px'
})), _defineProperty(_labelFontVariants, FontVariant.LabelSmallEmphasis, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontWeight: FontWeight.Medium,
  fontSize: '12px',
  lineHeight: '14px'
})), _defineProperty(_labelFontVariants, FontVariant.LabelSmallUp, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontSize: '13px',
  lineHeight: '16px'
})), _defineProperty(_labelFontVariants, FontVariant.LabelSmallUpEmphasis, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontWeight: FontWeight.Medium,
  fontSize: '13px',
  lineHeight: '16px'
})), _defineProperty(_labelFontVariants, FontVariant.LabelMedium, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontSize: '14px',
  lineHeight: '16px'
})), _defineProperty(_labelFontVariants, FontVariant.LabelMediumEmphasis, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontWeight: FontWeight.Medium,
  fontSize: '14px',
  lineHeight: '16px'
})), _defineProperty(_labelFontVariants, FontVariant.LabelLarge, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontSize: '16px',
  lineHeight: '18px'
})), _defineProperty(_labelFontVariants, FontVariant.LabelLargeEmphasis, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontWeight: FontWeight.Medium,
  fontSize: '16px',
  lineHeight: '18px'
})), _defineProperty(_labelFontVariants, FontVariant.LabelLargeUp, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontWeight: FontWeight.Regular,
  fontSize: '18px',
  lineHeight: '20px'
})), _defineProperty(_labelFontVariants, FontVariant.LabelExtraLargeEmphasis, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontWeight: FontWeight.Medium,
  fontSize: '24px',
  lineHeight: '28px'
})), _defineProperty(_labelFontVariants, FontVariant.LabelExtraLarge, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontWeight: FontWeight.Regular,
  fontSize: '24px',
  lineHeight: '28px'
})), _defineProperty(_labelFontVariants, FontVariant.LabelXExtraLargeEmphasis, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontWeight: FontWeight.Medium,
  fontSize: '32px',
  lineHeight: '36px'
})), _defineProperty(_labelFontVariants, FontVariant.LabelXExtraLarge, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontSize: '32px',
  lineHeight: '36px'
})), _defineProperty(_labelFontVariants, FontVariant.LabelXXExtraLargeEmphasis, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontWeight: FontWeight.Medium,
  fontSize: '48px',
  lineHeight: '54px'
})), _defineProperty(_labelFontVariants, FontVariant.LabelXXExtraLarge, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontSize: '48px',
  lineHeight: '54px'
})), _defineProperty(_labelFontVariants, FontVariant.LabelXXXExtraLarge, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontSize: '60px',
  lineHeight: '68px'
})), _labelFontVariants);
var textFontVariants = (_textFontVariants = {}, _defineProperty(_textFontVariants, FontVariant.TextExtraSmall, _objectSpread(_objectSpread({}, labelFontVariants[FontVariant.LabelExtraSmallUp]), {}, {
  lineHeight: '14px'
})), _defineProperty(_textFontVariants, FontVariant.TextSmall, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontSize: '12px',
  lineHeight: '18px'
})), _defineProperty(_textFontVariants, FontVariant.TextSmallUp, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontSize: '13px',
  lineHeight: '18px'
})), _defineProperty(_textFontVariants, FontVariant.TextMedium, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontSize: '14px',
  lineHeight: '20px'
})), _defineProperty(_textFontVariants, FontVariant.TextLarge, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontSize: '16px',
  lineHeight: '22px'
})), _defineProperty(_textFontVariants, FontVariant.TextExtraLarge, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontSize: '18px',
  lineHeight: '26px'
})), _defineProperty(_textFontVariants, FontVariant.TextXExtraLarge, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontSize: '28px',
  lineHeight: '40px'
})), _textFontVariants);
var buttonFontVariants = (_buttonFontVariants = {}, _defineProperty(_buttonFontVariants, FontVariant.ButtonSmEmphasis, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontWeight: FontWeight.Medium,
  fontSize: '12px',
  lineHeight: '14px'
})), _defineProperty(_buttonFontVariants, FontVariant.ButtonSm, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontSize: '12px',
  lineHeight: '14px'
})), _defineProperty(_buttonFontVariants, FontVariant.ButtonSmUpEmphasis, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontWeight: FontWeight.Medium,
  fontSize: '13px',
  lineHeight: '16px'
})), _defineProperty(_buttonFontVariants, FontVariant.ButtonSmUp, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontSize: '13px',
  lineHeight: '16px'
})), _defineProperty(_buttonFontVariants, FontVariant.ButtonMdEmphasis, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontWeight: FontWeight.Medium,
  fontSize: '14px',
  lineHeight: '16px'
})), _defineProperty(_buttonFontVariants, FontVariant.ButtonMd, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontSize: '14px',
  lineHeight: '16px'
})), _defineProperty(_buttonFontVariants, FontVariant.ButtonLgEmphasis, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontWeight: FontWeight.Medium,
  fontSize: '16px',
  lineHeight: '18px'
})), _defineProperty(_buttonFontVariants, FontVariant.ButtonLg, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontSize: '16px',
  lineHeight: '18px'
})), _defineProperty(_buttonFontVariants, FontVariant.ButtonLgUpEmphasis, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontWeight: FontWeight.Medium,
  fontSize: '18px',
  lineHeight: '20px'
})), _defineProperty(_buttonFontVariants, FontVariant.ButtonLgUp, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontSize: '18px',
  lineHeight: '20px'
})), _buttonFontVariants);
var linkFontVariants = (_linkFontVariants = {}, _defineProperty(_linkFontVariants, FontVariant.LinkSmall, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontSize: '12px',
  lineHeight: '14px'
})), _defineProperty(_linkFontVariants, FontVariant.LinkSmallEmphasis, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontWeight: FontWeight.Medium,
  fontSize: '12px',
  lineHeight: '14px'
})), _defineProperty(_linkFontVariants, FontVariant.LinkSmallUp, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontSize: '13px',
  lineHeight: '16px'
})), _defineProperty(_linkFontVariants, FontVariant.LinkSmallUpEmphasis, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontWeight: FontWeight.Medium,
  fontSize: '13px',
  lineHeight: '16px'
})), _defineProperty(_linkFontVariants, FontVariant.LinkMedium, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontSize: '14px',
  lineHeight: '16px'
})), _defineProperty(_linkFontVariants, FontVariant.LinkMediumEmphasis, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontWeight: FontWeight.Medium,
  fontSize: '14px',
  lineHeight: '16px'
})), _defineProperty(_linkFontVariants, FontVariant.LinkLarge, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontSize: '16px',
  lineHeight: '18px'
})), _defineProperty(_linkFontVariants, FontVariant.LinkLargeEmphasis, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontWeight: FontWeight.Medium,
  fontSize: '16px',
  lineHeight: '18px'
})), _defineProperty(_linkFontVariants, FontVariant.LinkLargeUp, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontSize: '18px',
  lineHeight: '20px'
})), _defineProperty(_linkFontVariants, FontVariant.LinkLargeUpEmphasis, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontWeight: FontWeight.Medium,
  fontSize: '18px',
  lineHeight: '20px'
})), _linkFontVariants);
var inputFontVariants = (_inputFontVariants = {}, _defineProperty(_inputFontVariants, FontVariant.InputSm, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontSize: '12px',
  lineHeight: '14px'
})), _defineProperty(_inputFontVariants, FontVariant.InputSmUp, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontSize: '13px',
  lineHeight: '16px'
})), _defineProperty(_inputFontVariants, FontVariant.InputMd, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontSize: '14px',
  lineHeight: '16px'
})), _defineProperty(_inputFontVariants, FontVariant.InputLg, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontSize: '16px',
  lineHeight: '18px'
})), _inputFontVariants);
var titleFontVariants = (_titleFontVariants = {}, _defineProperty(_titleFontVariants, FontVariant.TitleH6, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontWeight: FontWeight.Medium,
  fontSize: '12px',
  lineHeight: '18px'
})), _defineProperty(_titleFontVariants, FontVariant.TitleH5, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontWeight: FontWeight.Medium,
  fontSize: '14px',
  lineHeight: '21px'
})), _defineProperty(_titleFontVariants, FontVariant.TitleH4, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontWeight: FontWeight.Medium,
  fontSize: '16px',
  lineHeight: '24px'
})), _defineProperty(_titleFontVariants, FontVariant.TitleH3, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontWeight: FontWeight.Medium,
  fontSize: '18px',
  lineHeight: '27px'
})), _defineProperty(_titleFontVariants, FontVariant.TitleH2, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontWeight: FontWeight.Medium,
  fontSize: '20px',
  lineHeight: '30px'
})), _defineProperty(_titleFontVariants, FontVariant.TitleH1, _objectSpread(_objectSpread({}, defaultFontValue), {}, {
  fontWeight: FontWeight.Medium,
  fontSize: '28px',
  lineHeight: '42px'
})), _titleFontVariants);
export var defaultTypography = {
  // htmlFontSize: '14px',
  fontFamily: fontFamily,
  baseLineHeight: 1.375,
  font: _objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread({}, labelFontVariants), textFontVariants), buttonFontVariants), linkFontVariants), inputFontVariants), titleFontVariants)
};
