// @scaleflex/ui/core/tag-field/tag-field.styles.js
import styled, { css } from 'styled-components';
import { generateClassNames, applyDisplayNames } from '../../utils/functions';
import { Color as PColor } from '../../utils/types/palette';
import { BorderRadiusSize as BRSize } from '../../utils/types/shape';
import StyledLabel from '../label/label.styles';
import StyledFormHint from '../form-hint/form-hint.styles';
import { sizeTagFieldMixin, fontTagFieldMixin } from './tag-field.mixin';
import { Size } from './types';
var baseClassName = 'TagField';
var TagFieldRoot = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'root')
}).withConfig({
  componentId: "sc-10q17fj-0"
})(["", "{margin-top:4px;}", "{margin-bottom:4px;}"], StyledFormHint.FormHint, StyledLabel.Label);
var TagInputFieldWrapper = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'tagInputFieldWrapper')
}).withConfig({
  componentId: "sc-10q17fj-1"
})(["position:relative;width:fit-content;"]);
var TagFieldWrapper = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'tagFieldWrapper')
}).withConfig({
  componentId: "sc-10q17fj-2"
})(function (_ref) {
  var size = _ref.size,
    error = _ref.error,
    theme = _ref.theme;
  return css(["overflow:hidden;border:1px solid ", ";border-radius:2px;background:transparent;width:310px;", "{margin-top:4px;}", "{margin-bottom:4px;}", ""], error ? theme.palette[PColor.Error] : theme.palette[PColor.BackgroundPrimaryStateless], StyledFormHint.FormHint, StyledLabel.Label, sizeTagFieldMixin[size]);
});
var TagFieldLoader = /*#__PURE__*/styled.span.attrs({
  className: generateClassNames(baseClassName, 'loader')
}).withConfig({
  componentId: "sc-10q17fj-3"
})(["display:flex;margin-left:8px;"]);
var TagFieldListWrapper = /*#__PURE__*/styled.ul.attrs({
  className: generateClassNames(baseClassName, 'listWrapper')
}).withConfig({
  componentId: "sc-10q17fj-4"
})(function (_ref2) {
  var $loading = _ref2.$loading;
  return css(["display:inline-flex;align-items:flex-start;flex-wrap:wrap;margin:0;padding:0;list-style:none;width:100%;min-height:80px;", "{svg{animation:", ";}}"], TagFieldLoader, $loading ? 'spinner 1.2s linear infinite' : '1.2s');
});
var TagFieldInput = /*#__PURE__*/styled.input.attrs({
  className: generateClassNames(baseClassName, 'input')
}).withConfig({
  componentId: "sc-10q17fj-5"
})(function (_ref3) {
  var theme = _ref3.theme;
  return css(["background:transparent;border:none;width:100%;outline:none;color:", ";"], theme.palette[PColor.LinkHover]);
});
var TagFieldInputWrapper = /*#__PURE__*/styled.li.attrs({
  className: generateClassNames(baseClassName, 'inputWrapper')
}).withConfig({
  componentId: "sc-10q17fj-6"
})(function (_ref4) {
  var _ref4$size = _ref4.size,
    size = _ref4$size === void 0 ? Size.Md : _ref4$size,
    theme = _ref4.theme;
  return css(["background:none;flex-grow:1;padding:4px 0px;width:100px;", "{&::-webkit-input-placeholder{color:", ";", "}}"], TagFieldInput, theme.palette[PColor.TextPlaceholder], fontTagFieldMixin[size]);
});
var TagFieldBottom = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'tagFieldBottom')
}).withConfig({
  componentId: "sc-10q17fj-7"
})(function (_ref5) {
  var theme = _ref5.theme;
  return css(["border-radius:", ";display:flex;justify-content:space-between;align-items:flex-start;padding:0px;"], theme.shape.borderRadius[BRSize.Sm]);
});
var TagFieldButtonsWrapper = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'tagFieldButtonsWrapper')
}).withConfig({
  componentId: "sc-10q17fj-8"
})(function () {
  return css(["display:flex;align-items:center;gap:10px;"]);
});
var TagFieldCopyIcon = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'tagFieldCopyIcon')
}).withConfig({
  componentId: "sc-10q17fj-9"
})(["cursor:pointer;height:16px;"]);
var TagFieldSuggestionWrapper = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'suggestionWrapper')
}).withConfig({
  componentId: "sc-10q17fj-10"
})(["position:relative;margin-top:16px;"]);
var TagFieldSuggestionLabel = /*#__PURE__*/styled.label.attrs({
  className: generateClassNames(baseClassName, 'suggestionLabel')
}).withConfig({
  componentId: "sc-10q17fj-11"
})(function (_ref6) {
  var theme = _ref6.theme;
  return css(["display:flex;align-items:center;font-size:12px;padding:6px 0;color:", ";"], theme.palette[PColor.LinkHover]);
});
var TagFieldSuggestionIcon = /*#__PURE__*/styled.span.attrs({
  className: generateClassNames(baseClassName, 'Icon')
}).withConfig({
  componentId: "sc-10q17fj-12"
})(function () {
  return css(["display:inline-flex;margin-left:4px;"]);
});
var TagFieldSuggestionWrapperList = /*#__PURE__*/styled.ul.attrs({
  className: generateClassNames(baseClassName, 'suggestionWrapperList')
}).withConfig({
  componentId: "sc-10q17fj-13"
})(["display:inline-flex;flex-wrap:wrap;margin:0;padding:0;width:100%;list-style:none;"]);
var TagFieldSuggestionList = /*#__PURE__*/styled.li.attrs({
  className: generateClassNames(baseClassName, 'suggestionList')
}).withConfig({
  componentId: "sc-10q17fj-14"
})(function (_ref7) {
  var theme = _ref7.theme;
  return css(["position:relative;display:flex;align-items:center;margin:0 8px 8px 0;padding:4px 12px;border-radius:", ";border:1px dashed ", ";background:transparent;color:", ";line-height:16.4px;list-style:none;user-select:none;max-height:24px;"], theme.shape.borderRadius[BRSize.Sm], theme.palette[PColor.LinkPrimary], theme.palette[PColor.LinkPrimary]);
});
var Styled = applyDisplayNames({
  TagFieldRoot: TagFieldRoot,
  TagInputFieldWrapper: TagInputFieldWrapper,
  TagFieldWrapper: TagFieldWrapper,
  TagFieldListWrapper: TagFieldListWrapper,
  TagFieldInputWrapper: TagFieldInputWrapper,
  TagFieldInput: TagFieldInput,
  TagFieldSuggestionWrapper: TagFieldSuggestionWrapper,
  TagFieldSuggestionLabel: TagFieldSuggestionLabel,
  TagFieldSuggestionIcon: TagFieldSuggestionIcon,
  TagFieldSuggestionWrapperList: TagFieldSuggestionWrapperList,
  TagFieldSuggestionList: TagFieldSuggestionList,
  TagFieldLoader: TagFieldLoader,
  TagFieldBottom: TagFieldBottom,
  TagFieldButtonsWrapper: TagFieldButtonsWrapper,
  TagFieldCopyIcon: TagFieldCopyIcon
});
export default Styled;
