// @scaleflex/ui/core/autocomplete/autocomplete.styles.js
import styled, { css } from 'styled-components';
import CrossOutline from '@scaleflex/icons/cross-outline';
import StyledLabel from '../label/label.styles';
import StyledFormHint from '../form-hint/form-hint.styles';
import { generateClassNames, applyDisplayNames } from '../../utils/functions';
import { Color as PColor } from '../../utils/types/palette';
import StyledTag from '../tag/tag.styles';
var baseClassName = 'Autocomplete';
var Autocomplete = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'root')
}).withConfig({
  componentId: "sc-uyc6rp-0"
})(["", "{margin-top:4px;}", "{margin-bottom:4px;}", "{max-width:100%;}", "{white-space:nowrap;overflow:hidden;text-overflow:ellipsis;display:block;}"], StyledFormHint.FormHint, StyledLabel.Label, StyledTag.Tag, StyledTag.Label);
var TickIcon = /*#__PURE__*/styled.span.attrs({
  className: generateClassNames(baseClassName, 'tickIcon')
}).withConfig({
  componentId: "sc-uyc6rp-1"
})(function (_ref) {
  var palette = _ref.theme.palette;
  return css(["color:", ";"], palette[PColor.AccentStateless]);
});
var CrossIcon = /*#__PURE__*/styled(CrossOutline).withConfig({
  componentId: "sc-uyc6rp-2"
})(["", ""], function (_ref2) {
  var palette = _ref2.theme.palette;
  return css(["color:", ";"], palette[PColor.IconsSecondary]);
});
var AutocompleteContainer = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'container')
}).withConfig({
  componentId: "sc-uyc6rp-3"
})([""]);
var Styled = applyDisplayNames({
  Autocomplete: Autocomplete,
  AutocompleteContainer: AutocompleteContainer,
  TickIcon: TickIcon,
  CrossIcon: CrossIcon
});
export default Styled;
