// @scaleflex/ui/core/notification/notification.styles.js
import styled, { css } from 'styled-components';
import { generateClassNames, applyDisplayNames } from '../../utils/functions';
import { BorderRadiusSize as BRSize } from '../../utils/types/shape';
import { NotificationStatus } from './types';
import { backgroundColorMixin, messageColorMixin, messageWithIconMixin, messageFontMixin } from './notification.mixin';
import { FontVariant } from '../../utils/types/typography/font-variant';
var baseClassName = 'Notification';
var Icon = /*#__PURE__*/styled.span.attrs({
  className: generateClassNames(baseClassName, 'Icon')
}).withConfig({
  componentId: "sc-47wt1h-0"
})(function (_ref) {
  var _ref$removeBackground = _ref.removeBackground,
    removeBackground = _ref$removeBackground === void 0 ? true : _ref$removeBackground;
  return css(["margin-right:", "px;"], removeBackground ? '6' : '12');
});
var NotificationWrapper = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'root')
}).withConfig({
  componentId: "sc-47wt1h-1"
})(function (_ref2) {
  var _ref2$status = _ref2.status,
    status = _ref2$status === void 0 ? NotificationStatus.Info : _ref2$status,
    removeBackground = _ref2.removeBackground,
    hideIcon = _ref2.hideIcon,
    isTitle = _ref2.isTitle,
    theme = _ref2.theme;
  return css(["display:flex;align-items:flex-start;position:relative;padding:0px;gap:32px;width:fit-content;border-radius:", ";", " ", " ", " ", " ", ""], theme.shape.borderRadius[BRSize.Md], messageColorMixin[status], theme.typography.font[FontVariant.TextExtraSmall], !removeBackground && css(["", " ", ""], backgroundColorMixin[status], messageFontMixin[status]), !hideIcon && !removeBackground && css(["", ""], messageWithIconMixin[status]), removeBackground && isTitle && css(["", "{margin-top:1px;}"], Icon));
});
var Notification = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'Notification')
}).withConfig({
  componentId: "sc-47wt1h-2"
})(function (_ref3) {
  var theme = _ref3.theme;
  return css(["display:flex;padding:12px 16px;margin-right:64px;border-radius:", ";"], theme.shape.borderRadius[BRSize.Md]);
});
var MessageWrapper = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'MessageWrapper')
}).withConfig({
  componentId: "sc-47wt1h-3"
})(["display:flex;flex-direction:column;justify-content:center;align-items:flex-start;padding:0px;gap:8px;"]);
var Title = /*#__PURE__*/styled.span.attrs({
  className: generateClassNames(baseClassName, 'Title')
}).withConfig({
  componentId: "sc-47wt1h-4"
})(function (_ref4) {
  var theme = _ref4.theme;
  return css(["margin-bottom:4px;", ";"], theme.typography.font[FontVariant.LabelSmallUpEmphasis]);
});
var Message = /*#__PURE__*/styled.span.attrs({
  className: generateClassNames(baseClassName, 'Message')
}).withConfig({
  componentId: "sc-47wt1h-5"
})(["margin-right:32px;"]);
var Close = /*#__PURE__*/styled.span.attrs({
  className: generateClassNames(baseClassName, 'Close')
}).withConfig({
  componentId: "sc-47wt1h-6"
})(["display:flex;flex-direction:row;justify-content:center;align-items:center;gap:4px;padding:2px;position:absolute;top:2px;right:2px;cursor:pointer;"]);
var Styled = applyDisplayNames({
  NotificationWrapper: NotificationWrapper,
  Notification: Notification,
  Icon: Icon,
  MessageWrapper: MessageWrapper,
  Title: Title,
  Message: Message,
  Close: Close
});
export default Styled;
