// @scaleflex/ui/core/date-picker/date-picker.component.js
import _extends from "@babel/runtime/helpers/extends";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["value", "onChange", "autoSelectToday", "maxDate", "minDate", "position", "label", "hint", "popperOptions", "InputProps", "calendarStyles", "inputGroupProps", "fullWidth"];
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
import React, { useState, useRef, useEffect } from 'react';
import PT from 'prop-types';
import CalendarIcon from '@scaleflex/icons/calendar';
import { intrinsicComponent, objectValues } from '../../utils/functions';
import { propTypes as inputPropTypes } from '../input/input.component';
import Calendar from '../calendar';
import { propTypes as popperPropTypes } from '../popper/popper.component';
import Styled from './date-picker.styles';
import { isYearFormRegex } from '../calendar/calendar.utils';
import { Position } from '../popper/types';
var Datepicker = intrinsicComponent(function (_ref, ref) {
  var _isYearFormRegex$test;
  var _ref$value = _ref.value,
    value = _ref$value === void 0 ? '' : _ref$value,
    onChange = _ref.onChange,
    autoSelectToday = _ref.autoSelectToday,
    _ref$maxDate = _ref.maxDate,
    maxDate = _ref$maxDate === void 0 ? '' : _ref$maxDate,
    _ref$minDate = _ref.minDate,
    minDate = _ref$minDate === void 0 ? '' : _ref$minDate,
    position = _ref.position,
    label = _ref.label,
    hint = _ref.hint,
    popperOptions = _ref.popperOptions,
    InputPropsData = _ref.InputProps,
    calendarStyles = _ref.calendarStyles,
    inputGroupProps = _ref.inputGroupProps,
    fullWidth = _ref.fullWidth,
    rest = _objectWithoutProperties(_ref, _excluded);
  var _useState = useState(false),
    _useState2 = _slicedToArray(_useState, 2),
    open = _useState2[0],
    setOpen = _useState2[1];
  var _useState3 = useState(value),
    _useState4 = _slicedToArray(_useState3, 2),
    inputValue = _useState4[0],
    setInputValue = _useState4[1];
  var _useState5 = useState(true),
    _useState6 = _slicedToArray(_useState5, 2),
    showPlaceholder = _useState6[0],
    setShowPlaceholder = _useState6[1];
  var _useState7 = useState(false),
    _useState8 = _slicedToArray(_useState7, 2),
    isHovering = _useState8[0],
    setIsHovering = _useState8[1];
  var datePickerRef = useRef(null);
  var maxDateTimestamp = new Date(maxDate).getTime();
  var minDateTimestamp = new Date(minDate).getTime();
  var maxYear = new Date(inputValue).getFullYear();
  var isYearForm = (_isYearFormRegex$test = isYearFormRegex.test) === null || _isYearFormRegex$test === void 0 ? void 0 : _isYearFormRegex$test.call(isYearFormRegex, maxYear);
  var disabled = rest.disabled,
    readOnly = rest.readOnly;
  var handleOnChange = function handleOnChange(dateInputValue) {
    var dateInputTimestamp = new Date(dateInputValue).getTime();
    var isDisabledDate = dateInputTimestamp >= maxDateTimestamp || dateInputTimestamp <= minDateTimestamp;
    setInputValue(dateInputValue);
    if (onChange && !isDisabledDate && isYearForm) {
      onChange(dateInputValue);
    }
  };
  var handleCalendarIcon = function handleCalendarIcon() {
    if (!disabled && !readOnly) setOpen(!open);
  };
  var handlePlaceholder = function handlePlaceholder() {
    if (!disabled && !readOnly) setShowPlaceholder(false);
  };
  useEffect(function () {
    if (autoSelectToday) return;
    if (!value) setInputValue('');else setInputValue(value);
  }, [value]);
  return /*#__PURE__*/React.createElement(Styled.DatePicker, {
    ref: datePickerRef
  }, /*#__PURE__*/React.createElement(Styled.DatePickerInput, _extends({
    label: label,
    fullWidth: fullWidth,
    hint: hint,
    showPlaceholder: setShowPlaceholder,
    value: inputValue,
    isHovering: isHovering,
    onChange: function onChange(_ref2) {
      var currentTarget = _ref2.currentTarget;
      return handleOnChange(currentTarget.value);
    },
    inputProps: _objectSpread({
      iconEnd: function iconEnd(props) {
        return /*#__PURE__*/React.createElement(CalendarIcon, props);
      },
      iconClickEnd: handleCalendarIcon,
      type: 'date',
      max: '9999-12-31'
    }, InputPropsData || {}),
    ref: ref
  }, inputGroupProps, rest)), !inputValue && rest.placeholder && showPlaceholder && /*#__PURE__*/React.createElement(Styled.Placeholder, {
    onClick: handlePlaceholder,
    onMouseEnter: function onMouseEnter() {
      return setIsHovering(true);
    },
    onMouseLeave: function onMouseLeave() {
      return setIsHovering(false);
    },
    disabled: disabled,
    readOnly: readOnly
  }, rest.placeholder), /*#__PURE__*/React.createElement(Calendar, {
    value: inputValue,
    open: open,
    setOpen: setOpen,
    onChange: handleOnChange,
    maxDate: maxDate,
    minDate: minDate,
    autoSelectToday: autoSelectToday,
    anchorEl: datePickerRef.current,
    position: position || 'bottom-start',
    popperOptions: popperOptions,
    calendarStyles: calendarStyles
  }));
});
Datepicker.defaultProps = {
  calendarStyles: {}
};
export var propTypes = {
  value: PT.string,
  position: PT.oneOf(objectValues(Position)),
  maxDate: PT.string,
  minDate: PT.string,
  onChange: PT.func,
  label: PT.string,
  hint: PT.string,
  popperOptions: popperPropTypes.popperOptions,
  autoSelectToday: PT.bool,
  fullWidth: PT.bool,
  InputProps: PT.exact(inputPropTypes),
  inputGroupProps: PT.exact(inputPropTypes),
  calendarStyles: PT.object
};
Datepicker.propTypes = propTypes;
export default Datepicker;
