// @scaleflex/ui/core/gallery-button/gallery-button.styles.js
import styled, { css } from 'styled-components';
import { generateClassNames, applyDisplayNames } from '../../utils/functions';
import IconButtonStyled from '../icon-button/icon-button.styles';
import { Color as PColor } from '../../utils/types/palette';
var baseClassName = 'GalleryButton';
var Wrapper = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'wrapper')
}).withConfig({
  componentId: "sc-sshd3c-0"
})(["display:flex;justify-content:center;align-items:center;padding:4px 0px;gap:2px;"]);
var SideRectangle = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'side-rectangle')
}).withConfig({
  componentId: "sc-sshd3c-1"
})(function (_ref) {
  var theme = _ref.theme;
  return css(["width:6px;height:10px;border-radius:1.33px;background-color:", ";"], theme.palette[PColor.IconsPrimary]);
});
var MainRectangle = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'main-recangle')
}).withConfig({
  componentId: "sc-sshd3c-2"
})(function (_ref2) {
  var theme = _ref2.theme;
  return css(["width:12px;height:16px;border-radius:2.67px;background-color:", ";"], theme.palette[PColor.IconsSecondary]);
});
var GalleryButton = /*#__PURE__*/styled(IconButtonStyled.IconButton).attrs({
  className: generateClassNames(baseClassName, 'root')
}).withConfig({
  componentId: "sc-sshd3c-3"
})(function (_ref3) {
  var theme = _ref3.theme,
    active = _ref3.active;
  return css(["border:none;background-color:transparent;padding:8px 6px;&:hover{background-color:white;", "{background-color:", ";}", "{background-color:", ";}}&:active{background-color:", ";", "{background-color:", ";}", "{background-color:", ";}}&:focus{background-color:", ";border:none;", "{background-color:", ";}", "{background-color:", ";}}", " &:disabled{color:", ";background:", ";border:none;cursor:default;", "{background-color:", ";}", "{background-color:", ";}}"], MainRectangle, theme.palette[PColor.IconsPrimary], SideRectangle, theme.palette[PColor.IconsPrimaryHover], theme.palette[PColor.BackgroundActive], MainRectangle, theme.palette[PColor.LinkHover], SideRectangle, theme.palette[PColor.LinkActive], theme.palette[PColor.BackgroundActive], MainRectangle, theme.palette[PColor.LinkHover], SideRectangle, theme.palette[PColor.LinkActive], active && css(["background-color:", ";", "{background-color:", ";}", "{background-color:", ";}&:hover,&:active,&:focus{background-color:", ";", "{background-color:", ";}", "{background-color:", ";}}"], theme.palette[PColor.BackgroundActive], MainRectangle, theme.palette[PColor.LinkHover], SideRectangle, theme.palette[PColor.LinkActive], theme.palette[PColor.BackgroundActive], MainRectangle, theme.palette[PColor.LinkHover], SideRectangle, theme.palette[PColor.LinkActive]), theme.palette[PColor.BordersDisabled], theme.palette[PColor.Accent_1_2_Opacity], MainRectangle, theme.palette[PColor.IconsSecondary], SideRectangle, theme.palette[PColor.IconsPrimary]);
});
var Styled = applyDisplayNames({
  GalleryButton: GalleryButton,
  Wrapper: Wrapper,
  SideRectangle: SideRectangle,
  MainRectangle: MainRectangle
});
export default Styled;
