// @scaleflex/ui/theme/roots/breakpoints/entity/default-breakpoints.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
var _values, _classes;
import { Breakpoint } from '../../../../utils/types/css';
export var defaultBreakpoints = {
  keys: [Breakpoint.Xs, Breakpoint.Sm, Breakpoint.Md, Breakpoint.Lg, Breakpoint.Xl],
  values: (_values = {}, _defineProperty(_values, Breakpoint.Xs, 0), _defineProperty(_values, Breakpoint.Sm, 576), _defineProperty(_values, Breakpoint.Md, 768), _defineProperty(_values, Breakpoint.Lg, 1084), _defineProperty(_values, Breakpoint.Xl, 1342), _defineProperty(_values, Breakpoint.Xxl, 1600), _defineProperty(_values, Breakpoint.Xxxl, 1920), _values),
  classes: (_classes = {}, _defineProperty(_classes, Breakpoint.Xs, '.sfx-breakpoint-xs'), _defineProperty(_classes, Breakpoint.Sm, '.sfx-breakpoint-sm'), _defineProperty(_classes, Breakpoint.Md, '.sfx-breakpoint-md'), _defineProperty(_classes, Breakpoint.Lg, '.sfx-breakpoint-lg'), _defineProperty(_classes, Breakpoint.Xl, '.sfx-breakpoint-xl'), _defineProperty(_classes, Breakpoint.Xxl, '.sfx-breakpoint-xxl'), _defineProperty(_classes, Breakpoint.Xxxl, '.sfx-breakpoint-xxxl'), _classes),
  getBreakpointClass: function getBreakpointClass() {
    return '';
  },
  up: function up() {
    return '';
  },
  down: function down() {
    return '';
  },
  between: function between() {
    return '';
  },
  only: function only() {
    return '';
  }
};
